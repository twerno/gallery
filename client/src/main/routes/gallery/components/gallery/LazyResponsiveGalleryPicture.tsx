import AnimatedLoader from 'main/components/AnimatedLoader';
import { LazyLoaderTrigger } from 'main/components/LazyLoader';
import * as React from 'react';
import { GalleryResponsivePicture, IResponsivePictureProps } from './GalleryResponsivePicture';
import GalleryStyles from './styles/GalleryStyles';

export interface ILazyResponsivePictureProps {
    containerRef: React.RefObject<HTMLDivElement>;
    pictureData: IResponsivePictureProps
}

export const LazyResponsiveGalleryPicture = (props: ILazyResponsivePictureProps) => {
    return (
        <LazyLoaderTrigger
            loadingPlaceholder={loadingPlaceHolder}
            containerRef={props.containerRef}
        >
            {
                ({ setLoaded, isLoaded }) =>
                    <GalleryResponsivePicture
                        isVisible={isLoaded}
                        onLoadHandler={() => setLoaded(true)}
                        pictureData={props.pictureData}
                    />
            }
        </LazyLoaderTrigger>
    );
};

const loadingPlaceHolder = <GalleryStyles.ItemLoadingPlaceholder children={<AnimatedLoader />} />;