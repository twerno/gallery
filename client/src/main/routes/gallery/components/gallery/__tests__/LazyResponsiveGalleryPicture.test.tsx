import '@testing-library/jest-dom/extend-expect';
import { wait } from '@testing-library/react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import * as React from 'react';
import { componentBasicTests } from 'test/componentBasicTests';
import { LazyResponsiveGalleryPicture } from '../LazyResponsiveGalleryPicture';

let observerCallback: ((entries: IntersectionObserverEntry[]) => void) | null = null;
let observer: { observe: jest.Mock, unobserve: jest.Mock, disconnect: jest.Mock } | null = null;

const mockObserveFn = (callback: (entries: IntersectionObserverEntry[]) => void) => {
    observerCallback = callback;
    observer = {
        observe: jest.fn(),
        unobserve: jest.fn(),
        disconnect: jest.fn(),
    }
    return observer;
};

window.IntersectionObserver = jest.fn().mockImplementation(mockObserveFn);

describe('HomePage', () => {

    const props = {
        pictureData: {
            alt: 'alt',
            defaultImgSrc: 'img.jpg',
            responsiveSrcList: [{ src: 'img-big.jpg', media: '(min-width: 600px)' }]
        }
    }

    const newContainerRef = () => ({ current: document.createElement('div') });

    componentBasicTests(
        "LazyResponsiveGalleryPicture - not visible test - children should not be visible",
        <LazyResponsiveGalleryPicture {...props} containerRef={newContainerRef()} />
    );


    test('LazyResponsiveGalleryPicture - element is visible - children should be render', async () => {

        const containerRef = React.createRef<HTMLDivElement>();
        const wrapper = mount(
            <div ref={containerRef}>
                <LazyResponsiveGalleryPicture {...props} containerRef={containerRef} />
            </div>
        );

        // not null or undefined
        expect(observerCallback).toBeTruthy();
        expect(containerRef.current).toBeTruthy();
        await wait(() => {
            if (observerCallback && containerRef.current) {
                observerCallback([
                    getIntersectionObserverElementFrom(containerRef.current)
                ])
            }
        });

        const expected = toJson(wrapper);
        expect(expected).toMatchSnapshot();
    });

    test('LazyResponsiveGalleryPicture - disconnects an observer', async () => {

        const containerRef = React.createRef<HTMLDivElement>();
        const wrapper = mount(
            <div ref={containerRef}>
                <LazyResponsiveGalleryPicture {...props} containerRef={containerRef} />
            </div>
        );

        // not null or undefined
        expect(observerCallback).toBeTruthy();
        expect(containerRef.current).toBeTruthy();
        await wait(() => {
            if (observerCallback && containerRef.current) {
                observerCallback([
                    getIntersectionObserverElementFrom(containerRef.current)
                ])
            }
        });

        await wait(() => {
            if (observerCallback && containerRef.current) {
                observerCallback([
                    getIntersectionObserverElementFrom(containerRef.current)
                ])
            }
        });

        expect(observer?.disconnect).toBeCalledTimes(1);
    });

});

function getIntersectionObserverElementFrom(element: HTMLDivElement) {
    return {
        boundingClientRect: element.getBoundingClientRect(),
        intersectionRatio: 1,
        intersectionRect: element.getBoundingClientRect(),
        isIntersecting: true,
        time: 0,
        target: element,
        rootBounds: element.getBoundingClientRect(),
    };
}