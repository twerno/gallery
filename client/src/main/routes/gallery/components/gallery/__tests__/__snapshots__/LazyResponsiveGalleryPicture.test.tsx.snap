// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage [LazyResponsiveGalleryPicture - not visible test - children should not be visible] Component basic tests matches the snapshot 1`] = `
<Component
  containerRef={
    Object {
      "current": <div />,
    }
  }
  loadingPlaceholder={
    <styled.div>
      <styled.div />
    </styled.div>
  }
>
  <Component />
</Component>
`;

exports[`HomePage LazyResponsiveGalleryPicture - element is visible - children should be render 1`] = `
.c1 {
  background-color: rgba(0,0,0,0.05);
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.c1::before {
  border: .25rem solid rgba(0,0,0,0.3);
  -webkit-animation: 4s infinite spin linear;
  animation: 4s infinite spin linear;
}

.c1::before,
.c1::after {
  content: '';
  position: absolute;
  width: 2rem;
  height: 2rem;
  border-right-color: transparent;
  border-left-color: transparent;
  border-radius: 50%;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 500px;
  width: 100%;
  height: 100%;
}

@media only screen and (max-width:600px) {
  .c0 {
    min-width: 200px;
  }
}

<div>
  <Component
    containerRef={
      Object {
        "current": .c1 {
  background-color: rgba(0,0,0,0.05);
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.c1::before {
  border: .25rem solid rgba(0,0,0,0.3);
  -webkit-animation: 4s infinite spin linear;
  animation: 4s infinite spin linear;
}

.c1::before,
.c1::after {
  content: '';
  position: absolute;
  width: 2rem;
  height: 2rem;
  border-right-color: transparent;
  border-left-color: transparent;
  border-radius: 50%;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 500px;
  width: 100%;
  height: 100%;
}

.c2 {
  opacity: 1;
  position: relative;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
  min-width: 100%;
  position: absolute;
  opacity: 0;
}

.c3 {
  object-fit: cover;
  min-width: 100%;
  height: 100%;
}

@media only screen and (max-width:600px) {
  .c0 {
    min-width: 200px;
  }
}

<div>
          <div
            class="c0"
          >
            <div
              class="c1"
            />
          </div>
          <picture
            class="c2"
          >
            <source
              media="(min-width: 600px)"
              srcset="img-big.jpg"
            />
            <img
              alt="alt"
              class="c3"
              src="img.jpg"
            />
          </picture>
        </div>,
      }
    }
    pictureData={
      Object {
        "alt": "alt",
        "defaultImgSrc": "img.jpg",
        "responsiveSrcList": Array [
          Object {
            "media": "(min-width: 600px)",
            "src": "img-big.jpg",
          },
        ],
      }
    }
  >
    <Component
      containerRef={
        Object {
          "current": <div>
            <div
              class="c0"
            >
              <div
                class="c1"
              />
            </div>
            <picture
              class=""
            >
              <source
                media="(min-width: 600px)"
                srcset="img-big.jpg"
              />
              <img
                alt="alt"
                class=""
                src="img.jpg"
              />
            </picture>
          </div>,
        }
      }
      loadingPlaceholder={
        <styled.div>
          <styled.div />
        </styled.div>
      }
    >
      <styled.div>
        <div
          className="c0"
        >
          <styled.div>
            <div
              className="c1"
            />
          </styled.div>
        </div>
      </styled.div>
    </Component>
  </Component>
</div>
`;
